flang/lib/Lower/CharacterExpr.cpp

```diff
@@ -228,10 +228,12 @@ mlir::Type Fortran::lower::CharacterExprHelper::getSeqTy(
 
 mlir::Value
 Fortran::lower::CharacterExprHelper::createEmbox(const fir::CharBoxValue &box) {
-  // BoxChar require a reference.
-  auto kind = getCharacterType(box).getFKind();
-  auto boxCharType = fir::BoxCharType::get(builder.getContext(), kind);
-  auto refType = getReferenceType(box);
+  // BoxChar require a reference. Base CharBoxValue of CharArrayBoxValue
+  // are ok here (do not require a scalar type)
+  auto charTy = recoverCharacterType<false>(box.getBuffer().getType());
```

Comment:
That it's OK for the type to be an array type. There is an assert that fires otherwise. In the long-term I think we could just get rid of this when character array expressions are implemented, but for now I will just change this to `recoverCharacterType<false /* checkForScalar */>`.